crates/red_knot_python_semantic/src/types/property_tests.rs

```diff
@@ -0,0 +1,249 @@
+        Ty::KnownClassInstance(KnownClass::FunctionType),
+        Ty::KnownClassInstance(KnownClass::SpecialForm),
+        Ty::KnownClassInstance(KnownClass::TypeVar),
+        Ty::KnownClassInstance(KnownClass::TypeAliasType),
+        Ty::KnownClassInstance(KnownClass::NoDefaultType),
+        Ty::TypingLiteral,
+        Ty::BuiltinClassLiteral("str"),
+        Ty::BuiltinClassLiteral("int"),
+        Ty::BuiltinClassLiteral("bool"),
+        Ty::BuiltinClassLiteral("object"),
+    ])
+    .unwrap()
+    .clone()
+}
+
+/// Constructs an arbitrary type.
+///
+/// The `size` parameter controls the depth of the type tree. For example,
+/// a simple type like `int` has a size of 0, `Union[int, str]` has a size
+/// of 1, `tuple[int, Union[str, bytes]]` has a size of 2, etc.
+fn arbitrary_type(g: &mut Gen, size: u32) -> Ty {
+    if size == 0 {
+        arbitrary_core_type(g)
+    } else {
+        match u32::arbitrary(g) % 4 {
+            0 => arbitrary_core_type(g),
+            1 => Ty::Union(
+                (0..*g.choose(&[2, 3]).unwrap())
+                    .map(|_| arbitrary_type(g, size - 1))
+                    .collect(),
+            ),
+            2 => Ty::Tuple(
+                (0..*g.choose(&[0, 1, 2]).unwrap())
+                    .map(|_| arbitrary_type(g, size - 1))
+                    .collect(),
+            ),
+            3 => Ty::Intersection {
+                pos: (0..*g.choose(&[0, 1, 2]).unwrap())
+                    .map(|_| arbitrary_type(g, size - 1))
+                    .collect(),
+                neg: (0..*g.choose(&[0, 1, 2]).unwrap())
+                    .map(|_| arbitrary_type(g, size - 1))
+                    .collect(),
+            },
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl Arbitrary for Ty {
+    fn arbitrary(g: &mut Gen) -> Ty {
+        const MAX_SIZE: u32 = 2;
+        arbitrary_type(g, MAX_SIZE)
+    }
+
+    fn shrink(&self) -> Box<dyn Iterator<Item = Self>> {
+        // This is incredibly naive. We can do much better here by
+        // trying various subsets of the elements in unions, tuples,
+        // and intersections. For now, we only try to shrink by
+        // reducing unions/tuples/intersections to a single element.
+        match self.clone() {
+            Ty::Union(types) => Box::new(types.into_iter()),
+            Ty::Tuple(types) => Box::new(types.into_iter()),
+            Ty::Intersection { pos, neg } => Box::new(pos.into_iter().chain(neg)),
+            _ => Box::new(std::iter::empty()),
+        }
+    }
+}
+
+impl<'db> Type<'db> {
+    /// Checks if a type contains `Never` in its tree.
+    ///
+    /// This is currently needed since we currently don't propagate/simplify types containing
+    /// `Never`. For example, the type `tuple[int, Never]` is equivalent to `Never`. Similarly,
+    /// we simplify `Never` in unions, but we would not simplify `str | tuple[int, Never]`, so
+    /// we need to descent into unions and intersections as well.
```

Comment:
It's not said clearly here, but I think the implication is that we should add this simplification of types containing `Never` (which doesn't seem too complicated? I guess it would require mostly that we introduce a builder pattern for constructing tuple types also?), and once we've done that, this could go away?

Maybe this should be a TODO and/or a new GH issue?

Reply:
https://github.com/astral-sh/ruff/pull/14744